### DISCUSSION TITLE
I2C speed at 320 kHz instead of 400 kHz on the STM32F411

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#14267] [STM32 / Pyboard] [answered]
Hi,
Any idea why I get an I2C bus speed of 320 kHz instead of 400 kHz on the STM32F411?
I tested this on the nucleo board.
```
i2c = I2C(1, freq=400_000)
print(i2c)
```
```
I2C(1, scl=B6, sda=B7, freq=320000)
```

When testing this with the SoftI2C it shows 500 kHz (!?) but I prefer the hardware I2C.

```
i2c = SoftI2C(scl=Pin("B6"), sda=Pin("B7"), freq=400_000)
print(i2c)
```
```
SoftI2C(scl=B6, sda=B7, freq=500000)
```

PS: Is there any difference between the I2C of the pyb module and the one from the machine module?

**Comment by robert-hh** [comment, marked as answer]
If you look into the data sheet and the way, the I2C hardware is set up, it's clear that 400k cannot be achieved. The I2C device clock is 24MHz. According to the data sheet, getting to 400KHz requires a multiple of 10 MHz for the I2C clock. You could get that by changing the clock configuration of the device, e.g. by setting the clock to 80Mhz or trying 100Mhz. You have to change the value of:
`#define MICROPY_HW_CLK_PLLN (192)`
in mpconfigboard.h to either 160 for 80Mhz clock or 200 for a 100MHz clock. The latter is out of the spec, but could work.
Note that the actual clock at the bus depends as well from the rise time of the clock signal, determined by the time constant of all R's and C's at the bus. A slow rise time will lower the clock frequency.

**Reply by robert-hh** [reply]
It has to be 80MHz, resulting in a 20MHz base clock for I2C.

**Reply by robert-hh** [reply]
But that does not work either, because is breaks USB.

**Reply by robert-hh** [reply]
b.t.w.: When setting freq=400_000 for SoftI2C, the object reports 500_000, but the actual frequency at SCL is 250_000. With hardware I2C and low capacity at the pin, the actual frequency is exactly 320_000. With an RC time constant of ~10µs at SCL, the actual frequency drops to 86kHz, event if the print of the I2C object still tells 320kHz.
